inloop/statistics/views.py

```diff
@@ -0,0 +1,137 @@
+import json
+from collections import defaultdict
+from json import JSONDecodeError
+
+import django.db.models.functions as functions
+from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.db import models
+from django.http import JsonResponse
+from django.views import View
+from django.views.generic import TemplateView
+
+from inloop.solutions.models import Solution
+from inloop.tasks.models import Category, Task
+
+
+def get_truncator(
+    field_name: str, granularity: str
+) -> functions.datetime.TruncBase:
+    return {
+        "minute": functions.TruncMinute(field_name),
+        "hour": functions.TruncHour(field_name),
+        "day": functions.TruncDay(field_name),
+        "month": functions.TruncMonth(field_name),
+        "year": functions.TruncYear(field_name),
+    }[granularity]
+
+
+class FailureJsonResponse(JsonResponse):
+    reason = None
+
+    def __init__(self, *args, **kwargs):
+        if self.reason is None:
+            raise NotImplementedError("Failure JSON Responses must include a reason!")
+        super().__init__({
+            "reason": self.reason
+        }, *args, **kwargs)
```

Comment:
This looks weird. What about a small factory function?

```python
def bad_json_request(reason='Malformed JSON'):
    return JsonResponse({'reason': reason}, status=HTTPStatus.BAD_REQUEST)
```

Also note the use of the `http.HTTPStatus` enum (part of the stdlib).